### ISSUE TITLE
Create a control facility for installation/setup

### ISSUE THREAD

**Issue by jooola** [libretime/libretime#1528]
In order to better control the setup process with config file generation, database initial migration, I propose to write a python based CLI.

**Regarding the installation script:**

I previously proposed to rewrite the entire ./install script to python but by the time I wrote some parts of it I felt I was building something way too complex.
The complexity came because I wanted to make the installer configurable as much as possible, without using an orchestration tool (I proposed to use such orchestration tool but the idea was rejected as newcomers might have trouble with such tools). So I was basically creating an orchestration tool.

I don't want to replace the install script, but it should install a very common setup, with only a few configurable bits such as the listening port or running user.

For more complex setup, I propose to write a good documentation with code snippets, and/or the `ansible-role-libretime` I've been working on.

The installation process should only install the software, and not assume any configuration yet, as this should be handled in the setup part.

**Regarding the setup part of the installation (config file generation, database migration, storage directory creation):**

The current setup is in the legacy PHP code and I would like to extract this feature as cleanly as possible, but this might involve some big changes.

This setup part should handle the following:
- Prompt the user for the configuration details (with good default) and save the configuration to the libretime config file.
- Prompt the user whether the postgresql database/rabbitmq vhost should be created (default is no).
- Migrate the legacy database schema and default data.
- Configure and update the Icecast passwords in the icecast config file and the database (This was previously handled in both the install script and the php setup code, and this feels hacky).
- Create the storage directory and configure the music dir in the database.

Here are a few things that I am unsure:
- should the initial database migration have a dedicated command, and become a setup step ? In the future the api will have to run the django migration on each install/upgrade, this command would be a way to handle legacy and api migration until the legacy migration are rewritten as django migrations.
- should we move the legacy storage configuration from the database to the configuration file ? This makes the setup process less hacky and does not require use to create a 'MusicDir' during the setup process. And some parts of the configuration is already in the config file, so maybe we should merge everything together.
- #1451 seem to make sens, also regarding the icecast password setup above, which is hacky.
- ...

This step is part of #420 because the configuration file generation is a key part of the installation process.

I would like to get your thoughts about this, as while working on it, I noticed that it might actually be a big change.

**Comment by paddatrapper** [comment]
Why do you think this is the best way to go over something like a web-ui configuration wizard?

> Prompt the user whether the postgresql database/rabbitmq vhost should be created (default is no).

Default should probably be yes if we use the existing install script

**Comment by jooola** [comment]
> Why do you think this is the best way to go over something like a web-ui configuration wizard?

- We should not split the setup process in 2 different parts. As example, the install script prepare a password file for icecast, and the Web UI finalize those steps by updating some icecast password in the DB, which is not flexible at all, specially when you want to use other installation method. Another example is the database/rabbitmq settings, they are also passed using a file, and this makes using custom database/rabbitmq parameters difficult.
- The Web UI runs as the web user (www-data on debian based distros), and some task requires root in order to run some tasks, for example creating the storage directory, or to create a database (having a db user with CREATEDB role is not ideal, specially when working with a remote shared postgresql server).

Basically this does not allow for any flexibility when using something other than the install script.

In addition we want to get ride of the legacy code, so having a python cli based installer/setup tool will help us in anyway.

**Comment by paddatrapper** [comment]
> We should not split the setup process in 2 different parts. As example, the install script prepare a password file for icecast, and the Web UI finalize those steps by updating some icecast password in the DB, which is not flexible at all, specially when you want to use other installation method. Another example is the database/rabbitmq settings, they are also passed using a file, and this makes using custom database/rabbitmq parameters difficult.

This proposal still does though, it just changes where some of the configuration changes live from the install script to the control facility. With #1451, things are flexible for other configuration mechanisms regardless of the control facility.

> The Web UI runs as the web user (www-data on debian based distros), and some task requires root in order to run some tasks, for example creating the storage directory, or to create a database (having a db user with CREATEDB role is not ideal, specially when working with a remote shared postgresql server).

I agree that this is a problem

> In addition we want to get ride of the legacy code, so having a python cli based installer/setup tool will help us in anyway.

It could be a migration of the configuration wizard to the API/Vue-based UI instead of a dedicated Python application.

**Comment by jooola** [comment]
> This proposal still does though, it just changes where some of the configuration changes live from the install script to the control facility. With #1451, things are flexible for other configuration mechanisms regardless of the control facility.

It is split in 2 tools but running the control facility from the install script makes it a single step for random users installing Libretime. The control facility can be run by other installation tools for more advanced use cases.

This allow passing install details directly to the setup tool using env var or flags.

As opposed to the 2 step installation that require the use to first run the install script and then head to the ui.

> It could be a migration of the configuration wizard to the API/Vue-based UI instead of a dedicated Python application.

This would raise the same problem as the current install process, we have an install script and then we have to head to the web ui while the user running the webUI will not have enough privileges to run the extra steps.

---

I think it is easier to prompt all details as soon as possible in the installation process, build a solid configuration file, and run the installation from there.

If the concern is about inexperienced users trying to install Libretime, I don't think a web UI makes its easier. Solid documentation about the different install options and requirements should make it easier.

**Comment by jooola** [comment]
In addition to my comment on building a solid cli prompt to replace the UI forms, we could even use dialogs similar to the [whiptail](https://en.wikibooks.org/wiki/Bash_Shell_Scripting/Whiptail) dialog program. I already have a fancy cli prompt, but maybe full dialog screen make it even more documented/easy to use?

**Comment by jooola** [comment]
Also, instead of creating another new libretime tool, we could add all those commands to the libretime-api and run everything from there.

**Comment by jooola** [comment]
Closing this since the few commands we requires will be shipped a django commands.
